File: 2583-kth-largest-sum-in-a-binary-tree/2583-kth-largest-sum-in-a-binary-tree.java
class Solution {
    public long kthLargestLevelSum(TreeNode root, int k) {
        var maxHeap = new PriorityQueue<Long>((n1, n2) -> n2.compareTo(n1));
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while(!queue.isEmpty()){
            long curr = 0;
            for(var count=queue.size(); count>0; count--){
                var node = queue.poll();
                curr += node.val;
                if(node.left != null) queue.offer(node.left);
                if(node.right != null) queue.offer(node.right);
            }
            maxHeap.offer(curr);
        }

        while(k-- > 1 && !maxHeap.isEmpty()) maxHeap.poll();
        if(maxHeap.isEmpty()) return -1;
        return maxHeap.poll();
    }
}

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

File: 2583-kth-largest-sum-in-a-binary-tree/2583-kth-largest-sum-in-a-binary-tree.cs
public class Solution {
    public long KthLargestLevelSum(TreeNode root, int k) {
        var maxHeap = new PriorityQueue<long, long>();
        var queue = new Queue<TreeNode>();
        queue.Enqueue(root);

        while(queue.Count > 0){
            long curr = 0;
            for(var count=queue.Count; count>0; count--){
                var node = queue.Dequeue();
                curr += node.val;
                if(node.left != null) queue.Enqueue(node.left);
                if(node.right != null) queue.Enqueue(node.right);
            }
            maxHeap.Enqueue(curr, -curr);
        }

        while(k-- > 1 && maxHeap.Count > 0) maxHeap.Dequeue();
        if(maxHeap.Count == 0) return -1;
        return maxHeap.Dequeue();
    }
}

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */